GVFS/GVFS.Common/GitHubUpgrader.cs

```diff
@@ -523,7 +523,27 @@ public bool TryLoad(out string error)
                 this.UpgradeRing = RingType.NoConfig;
 
                 string ringConfig = null;
-                if (this.LocalConfig.TryGetConfig(GVFSConstants.LocalGVFSConfig.UpgradeRing, out ringConfig, out error))
+                string upgradeFeedUrl = null;
+
+                // 1. Try to read ring from upgrade.feedurl first.
+                if (this.LocalConfig.TryGetConfig(GVFSConstants.LocalGVFSConfig.UpgradeFeedUrl, out upgradeFeedUrl, out error) &&
+                    !string.IsNullOrEmpty(upgradeFeedUrl))
+                {
+                    string[] expectedRings = new string[] { RingType.Slow.ToString(), RingType.Fast.ToString() };
+                    foreach (string ring in expectedRings)
+                    {
+                        string ringSubstring = $"@{ring}/";
+                        if (upgradeFeedUrl.IndexOf(ringSubstring, StringComparison.OrdinalIgnoreCase) != -1)
+                        {
+                            ringConfig = ring;
+                            break;
+                        }
+                    }
+                }
+
+                // 2. If feedurl didn't specify a valid ring, then fallback on old behavior (look for upgrade.ring)

```

Comment:
Are there any valid reasons for the feedurl to not include a ring?

I'm wondering if we have to support falling back to the old config setting.

Reply:
cc: @jrbriggs and @jamill for their thoughts as well

Reply:
The Feed@View syntax is (as far as I know) an Azure DevOps thing. Falling back to the ring would allow a ringed deployment model on other nuget hosters.

Reply:
Ok thanks! 

And am I correct in assuming that someone could push official signed packages to their own feed? (I'm guessing that's one way we could get in this scenario?)

Reply:
Yeah. Admittedly it's a bit of a contrived scenario. I wouldn't be opposed to simplifying here and picking one way or another. I've gone back and forth on this several times.

Reply:
> The Feed@View syntax is (as far as I know) an Azure DevOps thing. Falling back to the ring would allow a ringed deployment model on other nuget hosters.

Actually, if this is an Azure DevOps thing, @alameenshah why is this check happening in `GitHubUpgraderConfig`?  Is it because the GitHub upgrader is what's used to check for new versions?

(If this is the case, it would be helpful to include that detail in the comments)

Reply:
> Yeah. Admittedly it's a bit of a contrived scenario. I wouldn't be opposed to simplifying here and picking one way or another. I've gone back and forth on this several times.

@jrbriggs, @alameenshah  another option (and I'm going back and forth on how I feel about it) is that the ring setting, if present, will override whatever is picked up automatically from the feed name.  

Some potential advantages:

- A ring config of "None" will disable auto-upgrade (even if the feed has a ring name in it).

- If someone has a different naming scheme in their feed (e.g. "Lab", "Fast", "Canary") they can manually control which GitHub release should drive the auto-upgrade notifications.

> Admittedly it's a bit of a contrived scenario

My scenario might be even *more* contrived, but please let me know what you think.

Reply:
For this initial release, let's keep this as simple as possible and treat these non-mainline scenarios as out-of-scope. I would rather explicitly treat these as not-handled / not-specified, rather than semi-support them and get locked into a behavior we didn't think through (or give the hint they are supported and then change the behavior in the future).

If `upgrade.feedurl` is specified, then we assume that the user intended to use `NuGetUpgrader`. If we cannot find a corresponding GitHub release ring, then prompt for auto-upgrade is disabled.

If the ring setting is present in addition to a feedurl, then we can either ignore it, or treat it as a misconfiguration (as we are not sure of the user's intent). I don't think we should fall back to GitHubUpgrader in this case, as a misconfigured NuGetUpgrader could then result in the user unexpectedly using the GitHubUpgrader.

> * A ring config of "None" will disable auto-upgrade (even if the feed has a ring name in it).

Now - this might be a state we want to support. Do we need to support a state where a NuGet upgrader is configured, but the prompt to upgrade is disabled?

Reply:
As discussed, keeping it simple for the time being. (No fallback to using `upgrade.ring` when `feedURL` does not contain `Fast`/`Slow` rings info.)